src/node_native_module.cc

```diff
@@ -174,6 +174,48 @@ MaybeLocal<Function> NativeModuleLoader::CompileAsModule(
   return LookupAndCompile(context, id, &parameters, result);
 }
 
+Local<String> NativeModuleLoader::LoadInternalModuleSource(Isolate* isolate,
+                                                           const char* id) {
+#ifdef NODE_LOAD_INTERNAL_MODULES_FROM_DISK
+    // Filename on disk
+    std::string filename = "./lib/";
+    filename += id;
+    filename += ".js";
+
+    uv_fs_t req;
+    uv_file file =
+        uv_fs_open(nullptr, &req, filename.c_str(), O_RDONLY, 0, nullptr);
+    uv_fs_req_cleanup(&req);
+    if (req.result < 0) {
+      return {};
+    }
+
+    std::string contents;
+    char buffer[4096];
+    uv_buf_t buf = uv_buf_init(buffer, sizeof(buffer));
+
+    while (true) {
+      const int r =
+          uv_fs_read(nullptr, &req, file, &buf, 1, contents.length(), nullptr);
+      uv_fs_req_cleanup(&req);
+      if (r <= 0) {
+        break;
+      }
+      contents.append(buf.base, r);
+    }
+
+    return String::NewFromUtf8(isolate,
+                               contents.c_str(),
+                               v8::NewStringType::kNormal,
+                               contents.length())
+        .ToLocalChecked();
```

Comment:
Rule Of Three: there are now three places with very similar logic (InternalModuleReadJSON in node_file.cc, ReadFile in module_wrap.cc and now this function and file) so it's time to refactor.

InternalModuleReadJSON is arguably the most optimal implementation (e.g. doesn't do a final superfluous zero read like ReadFile and this function do) so I'd base it on that.

Reply:
yeah I copied this from module_wrap. where should I put the common function?

Reply:
util.cc? That's already a grab bag of unrelated stuff. :-)

Reply:
I'm actually going to modify this to throw V8 exceptions, so I don't think I can factor it out.